src/toil/lib/plugins.py

```diff
@@ -0,0 +1,66 @@
+import importlib
+from typing import Any, Literal
+import pkgutil
+from toil.lib.memoize import memoize
+
+
+PluginType = Literal["batch_system"] | Literal["url_access"]
+plugin_types: list[PluginType] = ["batch_system", "url_access"]
+
+_registry: dict[str, dict[str, Any]] = {k: {} for k in plugin_types}
+
+def register_plugin(
+    plugin_type: PluginType, plugin_name: str, plugin_being_registered: Any
+) -> None:
+    """
+    Adds a plugin to the registry for the given type of plugin.
+    """
+    _registry[plugin_type][plugin_name] = plugin_being_registered
+
+def remove_plugin(
+    plugin_type: PluginType, plugin_name: str) -> None:
+    """
+    Removes a plugin from the registry for the given type of plugin.
+    """
+    try:
+        del _registry[plugin_type][plugin_name]
+    except KeyError:
+        # If the plugin does not exist, it can be ignored
+        pass
+
+def get_plugin_names(plugin_type:PluginType) -> list[str]:
+    """
+    Get the names of all the available plugins.
+    """
+    _load_all_plugins(plugin_type)
+    return list(_registry[plugin_type].keys())
+
+def get_plugin(plugin_type: PluginType, plugin_name: str) -> Any:
+    """
+    Get a plugin class by name.
```

Comment:
This sounds a little misleading to me because when I read it I think that:
1. The result will be a class, when it's really a function that returns a class when called.
2. The argument is the name of the class (like `toil_plugin_url_access_foobar.lib.plugin.MyCoolClass`), when it's really a user-facing batch system name or URL scheme.